lib/astc_encode.cpp

```diff
@@ -0,0 +1,829 @@
+        pthread_join(threadDescs[i].threadHandle, nullptr);
+    }
+
+    delete[] threadDescs;
+}
+
+/**
+ * @memberof ktxTexture
+ * @ingroup writer
+ * @~English
+ * @brief Encode and compress a ktx texture with uncompressed images to astc.
+ *
+ * The images are encoded to ASTC block-compressed format. The encoded images
+ * replace the original images and the texture's fields including the DFD are
+ * modified to reflect the new state.
+ *
+ * Such textures can be directly uploaded to a GPU via a graphics API.
+ *
+ * @param[in]   This   pointer to the ktxTexture object of interest.
+ * @param[in]   params pointer to ASTC params object.
+ *
+ * @return      KTX_SUCCESS on success, other KTX_* enum values on error.
+ *
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture's images are supercompressed.
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture's images are in a block compressed
+ *                              format.
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture image's format is a packed format
+ *                              (e.g. RGB565).
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture image format's component size is not
+ *                              8-bits.
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture's images are 1D. Only 2D images can
+ *                              be supercompressed.
+ * @exception KTX_INVALID_OPERATION
+ *                              ASTC  compressor failed to compress image for any
+                                reason.
+ * @exception KTX_OUT_OF_MEMORY Not enough memory to carry out compression.
+ */
+extern "C" KTX_error_code
+ktxTexture_CompressAstcEx(ktxTexture* _This, ktxAstcParams* params) {
+    // FIXME: At the moment defaults to ktx2 textures only
+    assert(_This->classId == ktxTexture2_c && "Only support ktx2 ASTC.");
+    ktxTexture2* This = (ktxTexture2*)_This;
+
+    KTX_error_code result;
+
+    if (!params)
+        return KTX_INVALID_VALUE;
+
+    if (params->structSize != sizeof(struct ktxAstcParams))
+        return KTX_INVALID_VALUE;
+
+    if (This->supercompressionScheme != KTX_SS_NONE)
+        return KTX_INVALID_OPERATION; // Can't apply multiple schemes.
+
+    if (This->isCompressed)
+        return KTX_INVALID_OPERATION;  // Only non-block compressed formats
+                                       // can be encoded into an ASTC format.
+
+    if (This->_protected->_formatSize.flags & KTX_FORMAT_SIZE_PACKED_BIT)
+        return KTX_INVALID_OPERATION;
+
+    // Basic descriptor block begins after the total size field.
+    const uint32_t* BDB = This->pDfd+1;
+
+    uint32_t num_components, component_size;
+    getDFDComponentInfoUnpacked(This->pDfd, &num_components, &component_size);
+
+    if (component_size != 1)
+        return KTX_INVALID_OPERATION; // Can only deal with 8-bit components at the moment
```

Comment:
This is wrong. This encoder can handle 16-bit components as well.

Reply:
This was removed when I wanted to add HDR support as well, but then we decided to do that in a separate PR and I [reverted](https://github.com/KhronosGroup/KTX-Software/pull/433#discussion_r649100986) it and will be removed when we do add HDR/16bit image support.

Reply:
I was thinking of 16-bit PNGs as input. If there is no HDR encoding support yet then those do not matter.